Documentation/using-dex.md

```diff
@@ -0,0 +1,188 @@
+
+```yaml
+staticClients:
+- id: example-app
+  secret: example-app-secret
+  name: 'Example App'
+  # Where the app will be running.
+  redirectURIs:
+  - 'http://127.0.0.1:5555/callback'
+```
+
+In this case, the Go code would configured as:
+
+```go
+// Initialize a provider by specifying dex's issuer URL.
+provider, err := oidc.NewProvider(ctx, "https://dex-issuer-url.com")
+if err != nil {
+    // handle error
+}
+
+// Configure the OAuth2 config with the client values.
+oauth2Config := oauth2.Config{
+    // client_id and client_secret of the client.
+    ClientID:     "example-app",
+    ClientSecret: "example-app-secret",
+
+    // The redirectURL.
+    RedirectURL: "http://127.0.0.1:5556/callback",
+
+    // Discovery returns the OAuth2 endpoints.
+    Endpoint: provider.Endpoint(),
+
+    // "openid" is a required scope for OpenID Connect flows.
+    //
+    // Other scopes, such as "groups" can be requested.
+    Scopes: []string{oidc.ScopeOpenID, "profile", "email", "groups"},
+}
+
+// Create an ID token parser.
+idTokenVerifier := provider.NewVerifier(&oidc.Config{ClientID: "example-app"})
+```
+
+The HTTP server should then redirect unauthenticated users to dex to initialize the OAuth2 flow.
+
+```go
+// handleRedirect is used to start an OAuth2 flow with the dex server.
+func handleRedirect(w http.ResponseWriter, r *http.Request) {
+    state := newState()
+    http.Redirect(w, r, oauth2Config.AuthCodeURL(state), http.StatusFound)
+}
+```
+
+After dex verifies the user's identity it redirects the user back to the client app with a code that can be exchanged for an ID token. The ID token can then be parsed by the verifier created above. This immediately 
+
+```go
+func handleOAuth2Callback(w http.ResponseWriter, r *http.Request) {
+    state := r.URL.Query().Get("state")
+
+    // Verify state.
+
+    oauth2Token, err := oauth2Config.Exchange(ctx, r.URL.Query().Get("code"))
+    if err != nil {
+        // handle error
+    }
+
+    // Extract the ID Token from OAuth2 token.
+    rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+    if !ok {
+        // handle missing token
+    }
+
+    // Parse and verify ID Token payload.
+    idToken, err := idTokenVerifier.Verify(ctx, rawIDToken)
+    if err != nil {
+        // handle error
+    }
+
+    // Extract custom claims.
+    var claims struct {
+        Email    string   `json:"email"`
+        Verified bool     `json:"email_verified"`
+        Groups   []string `json:"groups"`
+    }
+    if err := idToken.Claims(&claims); err != nil {
+        // handle error
+    }
+}
+```
+
+### State tokens
+
+The state parameter is an arbitrary string that dex will always return with the callback. It plays a security role, preventing certain kinds of OAuth2 attacks. Specifically it can be used by clients to ensure:
+
+* The user who started the flow is the one who finished it by cooking the user with some data that can be associated with the state.
```

Comment:
> The use who started the flow is the one who finished it. This can be accomplished by cooking the user with some data that can be associated with the state (possibly the entire state token itself).

That work?